politeiawww/politeiawww.go

```diff
@@ -858,6 +866,78 @@ func (p *politeiawww) handleWebsocketWrite(wc *wsContext) {
 	}
 }
 
+// updateBestBlock updates the cached best block.
+func (p *politeiawww) updateBestBlock(bestBlock uint64) {
+	p.bbMtx.Lock()
+	defer p.bbMtx.Unlock()
+	p.bestBlock = bestBlock
+}
+
+// getBestBlock returns the cached best block if there is an active websocket
+// connection to dcrdata. Otherwise, it requests the best block from politeiad
+// using the the decred plugin best block command.
+func (p *politeiawww) getBestBlock() (uint64, error) {
+	p.bbMtx.RLock()
+	bb := p.bestBlock
+	p.bbMtx.RUnlock()
+
+	// the cached best block will equal 0 if no messages have been received
+	// since the startup of the service or since reconnection to wsDcrdata.
+	if bb == 0 || !p.wsDcrdata.isSubscribed(newBlockSub) {
+		return p.getBestBlockDecredPlugin()
+	}
+
+	return bb, nil
+}
+
+// resetPiDcrdataWSSubs is responsible for resetting the wsDcrdata connection
+// and making necessary changes to the required changes to the politeiawww
+// state so that the service continues to function properly during and after
+// the reconnection.
+func (p *politeiawww) resetPiDcrdataWSSubs() {
```

Comment:
Why do you need this?  

`if bb == 0 || !p.wsDcrdata.isSubscribed(newBlockSub) {`

This already makes sure that the decred plugin command is used when there is no subscription.

Reply:
There might be a subscription, but if no messages have been received yet, the cached value will be incorrect.

Reply:
If no messages have been received yet, won't bb be 0 and it will use the decred plugin command?

Reply:
Are you asking why I'm resetting the cached value to 0? 

This is because if we are disconnected from dcrdata and this value is not set to zero, then in the time after we reconnect and before we receive the first message, the value from the last message before we were disconnected will be returned to the user.

Reply:
Oh I see.  In that case, can't this line in `getBestBlock()`

`if bb == 0 || !p.wsDcrdata.isSubscribed(newBlockSub) {`

just be 

`if bb == 0 {` 

Since bb will be 0 when either:
1. The subscription is inactive
Or
2. The subscription is active but has not received the best block yet

Reply:
Yep that works.